### DISCUSSION TITLE
Migration problems pi4j v1.4 to 2.1.1

### DISCUSSION THREAD

**Discussion by StefanSchubert** [Pi4J/pi4j#203] [Q&A]
Hi there,

I'm trying to upgrade my noice threshold sensor (as KY-038 module from AZDelivery used with a  raspi 3 model b, rev1.2).
My setup (java based) works well with p4j v1 (using wiring pi pin 2).
Now I'm trying the v2 setup, where wiring 2 is mapped on GPIO/BCM 27

It all starts up well but my event listener does not fire at all.

On the pi I checked also via python with this:

```
#!/usr/bin/python
import RPi.GPIO as GPIO
import time
GPIO.setmode(GPIO.BCM)
GPIO.setup(27,GPIO.IN,pull_up_down=GPIO.PUD_DOWN)
while True:
  i = GPIO.input(27)
  if i != 0:
     print "GOT SIGNAL"
     print i
     break;

```

This works. As soon as I clap in my hand I got the "1" and the loop stops. Why does the new java api is not working?

used dependencies from pom.xml

```
        <!-- include Pi4J Core -->
        <dependency>
            <groupId>com.pi4j</groupId>
            <artifactId>pi4j-core</artifactId>
            <version>2.1.1</version>
        </dependency>

        <!-- include Pi4J Plugins (Platforms and I/O Providers) -->
        <dependency>
            <groupId>com.pi4j</groupId>
            <artifactId>pi4j-plugin-raspberrypi</artifactId>
            <version>2.1.1</version>
        </dependency>
        <dependency>
            <groupId>com.pi4j</groupId>
            <artifactId>pi4j-plugin-pigpio</artifactId>
            <version>2.1.1</version>
        </dependency>

```

And heres how I setup the PIN and listener (SENSOR_BCM_PIN = 27) :

```
        pi4jContext = Pi4J.newAutoContext();
        ....
            DigitalInputConfig noiceSensorConfig = DigitalInput.newConfigBuilder(pi4jContext)
                    .id("BCM-#"+SENSOR_BCM_PIN)
                    .address(SENSOR_BCM_PIN)
                    .name("KY-038 Module")
                    .description("DoorBellSound Detection using KY-038 module from AZ-Delivery in threshold mode")
                    .debounce(DEBOUNCE_DELAY_IN_SECS, TimeUnit.MILLISECONDS) // convert secs to millis
                    .pull(PullResistance.PULL_DOWN)
                    .provider("pigpio-digital-input")
                    .build();
           ....
            sensorInput = pi4jContext.din().create(noiceSensorConfig);
            sensorInput.addListener(pinEventListener);

          (... And the Listener ... )
          
          public class PinEventListener implements DigitalStateChangeListener {

                @Override
                public void onDigitalStateChange(DigitalStateChangeEvent event) {
            ....

```

pigpiod -v
79

Thanks for any help.
Stefan

**Comment by StefanSchubert** [comment]
As this is simply teasing me I'm still investigating.

I tried ~/pigpio-master/EXAMPLES/Python/GPIO_STATUS/gpio_status.py 

which complained that the deamon wasn't running. I'm not sure if the pi4jv2 requires that the deamon is running so I did a

```
sudo systemctl start pigpiod
sudo systemctl enable pigpiod
```

My EventListener is still not triggered, but I observed that gpio_status.py displays different values on pin 27 when I snap with my fingers near the sensor:

                       Status of gpios 0-31

 0=1  INPUT: 0           11=0  INPUT: 0           22=0  INPUT: 0
 1=1  INPUT: 0           12=0  INPUT: 0           23=0  INPUT: 0
 2=1  INPUT: 0           13=0  INPUT: 0           24=0  INPUT: 0
 3=1  INPUT: 0           14=0  INPUT: 0           25=0  INPUT: 0
 4=1  INPUT: 0           15=1  INPUT: 0           26=0  INPUT: 0
 5=1  INPUT: 0           16=0  INPUT: 0           **27=0  INPUT: 6**
 6=1  INPUT: 0           17=0  INPUT: 0           28=0  INPUT: 0
 7=1  INPUT: 0           18=0  INPUT: 0           29=1  INPUT: 0
 8=1  INPUT: 0           19=0  INPUT: 0           30=0  INPUT: 0
 9=0  INPUT: 0           20=0  INPUT: 0           31=0  INPUT: 0
10=0  INPUT: 0           21=0  INPUT: 0
 
and another try:

 0=1  INPUT: 0           11=0  INPUT: 0           22=0  INPUT: 0
 1=1  INPUT: 0           12=0  INPUT: 0           23=0  INPUT: 0
 2=1  INPUT: 0           13=0  INPUT: 0           24=0  INPUT: 0
 3=1  INPUT: 0           14=0  INPUT: 0           25=0  INPUT: 0
 4=1  INPUT: 0           15=1  INPUT: 0           26=0  INPUT: 0
 5=1  INPUT: 0           16=0  INPUT: 0           **27=0  INPUT: 42**
 6=1  INPUT: 0           17=0  INPUT: 0           28=0  INPUT: 0
 7=1  INPUT: 0           18=0  INPUT: 0           29=1  INPUT: 0
 8=1  INPUT: 0           19=0  INPUT: 0           30=0  INPUT: 0
 9=0  INPUT: 0           20=0  INPUT: 0           31=0  INPUT: 0
10=0  INPUT: 0           21=0  INPUT: 0

I thought that I would only get 0/1 here but it seems that I get the loudness?
Could it be that it requires a certain value such that pi4j signals an event on that PIN?

**Comment by StefanSchubert** [comment]
I also found out that I missed some preconditions:

pi4j was not available on the pi (though I think it's not necessary since it should be available within java through the maven dependencies) 

While https://pi4j.com/getting-started/set-up-a-new-raspberry-pi/ is missing the point of installing PI4J
I followed  https://github.com/Pi4J/pi4j.github.io/blob/main/docs/v2/install

so the pi shows


```
$ pi4j --version

--------------------------------------------
            THE Pi4J PROJECT
           (https://pi4j.com)
--------------------------------------------
   PI4J.VERSION     : 2.0
   PI4J.TIMESTAMP   : 2021-07-14 01:09:48
--------------------------------------------
```
However the java app, i.e. my Listener is not impressed by the change. Which part went wrong?

**Comment by FDelporte** [comment]
You're correct, installing Pi4J is not needed as PiGpio is compiled into the library, also the daemon is not used. 
So what could be wrong... maybe let's first check if the libraries are compiled correctly, as Java-modules can be challenging. 
For instance, when compiling to a FAT jar, specific settings are needed as is described on https://pi4j.com/documentation/building/fat-jar/

Can you add similar PrintInfo code from https://github.com/Pi4J/pi4j-example-minimal/blob/main/src/main/java/com/pi4j/example/MinimalExample.java#L101 to validate that everything is loaded correctly?

Ah, and do you start your application with `sudo`?

**Reply by StefanSchubert** [reply]
I will see to provide the information - will be tomorrow it's getting late now ;-)
I also just stumbled about the sudo is required for accessing pigpio and tried some things.

The fat jar will be started via systemd, I'm using this service file:
```

[Unit]
Description=Bell Range Extender Service
After=network-online.target

[Service]
User=pibell
Restart=on-failure
ExecStart=/var/pibell/pibell-service.jar

[Install]
WantedBy=multi-user.target

```
Also I added the pibell user in visudo with this

```
pibell  ALL=/var/pibell/pibell-service.jar
pibell  ALL=/bin/bash /var/pibell/pibell-service.jar

```
However I just stopped the service and restarted it from CLI with sudo...no effect :-(
Let's see tomorrow what the PrintInfo will show.

**Comment by FDelporte** [comment]
Just a hint when starting java with systemd: create a bash script to start the application you can execute successfully yourself, and use that in systemd file, eg `ExecStart=bash /var/pibell/start.sh`

**Reply by StefanSchubert** [reply]
I see, but what would be the advantage, when building already a self executable jar?
Currently I'm using ansible for rollout on the pi which manages all the stop service, demaon reload steps.
To check if the jar it self is working I just launch it from CLI.

**Comment by StefanSchubert** [comment]
OK two things I discovered in the meantime after checking your hints, especially:
_specific settings are needed as is described on https://pi4j.com/documentation/building/fat-jar/_

I obviously missed the shading part, which I added. The other difference I observed is, that I'm using springs plugin to build the self executable, instead of the maven jar plugin:

```
            <plugin>
                <groupId>org.springframework.boot</groupId>
                <artifactId>spring-boot-maven-plugin</artifactId>
                <configuration>
                    <!-- Build a fully executable jar to use it via /etc/init.d
                         see https://docs.spring.io/spring-boot/docs/current/reference/html/deployment-install.html
                    -->
                    <executable>true</executable>
                    <!-- Disabled Layers, as we have pi and not docker as target environment -->
                    <!--                <layers>-->
                    <!--                    <enabled>true</enabled>-->
                    <!--
                                </layers>
                    -->
                </configuration>
                <executions>
                    <execution>
                        <goals>
                            <goal>build-info</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
```

After adding the Shade Plugin and also integrating the PrintInfo I got the following on console:

```
Mar 11 10:05:21 MediaParkPiBell pibell-service.jar[2215]: PLATFORMS: [1] "Pi4J Runtime Platforms" <com.pi4j.platform.impl.DefaultPlatforms>
Mar 11 10:05:21 MediaParkPiBell pibell-service.jar[2215]: └─PLATFORM: "RaspberryPi Platform" {raspberrypi} <com.pi4j.plugin.raspberrypi.platform.RaspberryPiPlatform> {Pi4J Platform for the RaspberryPi series of products.}
Mar 11 10:05:21 MediaParkPiBell pibell-service.jar[2215]: PLATFORM: "RaspberryPi Platform" {raspberrypi} <com.pi4j.plugin.raspberrypi.platform.RaspberryPiPlatform> {Pi4J Platform for the RaspberryPi series of products.}
Mar 11 10:05:21 MediaParkPiBell pibell-service.jar[2215]: PROVIDERS: [12] "I/O Providers" <com.pi4j.provider.impl.DefaultProviders>
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: ├─ANALOG_OUTPUT: [0] <com.pi4j.io.gpio.analog.AnalogOutputProvider>
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: ├─I2C: [2] <com.pi4j.io.i2c.I2CProvider>
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: │ ├─PROVIDER: "PiGpio I2C Provider" {pigpio-i2c} <com.pi4j.plugin.pigpio.provider.i2c.PiGpioI2CProviderImpl> {com.pi4j.plugin.pigpio.provider.i2c.PiGpioI2CProviderImpl}
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: │ └─PROVIDER: "RaspberryPi I2C Provider" {raspberrypi-i2c} <com.pi4j.plugin.raspberrypi.provider.i2c.RpiI2CProviderImpl> {com.pi4j.plugin.raspberrypi.provider.i2c.RpiI2CProviderImpl}
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: ├─ANALOG_INPUT: [0] <com.pi4j.io.gpio.analog.AnalogInputProvider>
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: ├─PWM: [2] <com.pi4j.io.pwm.PwmProvider>
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: │ ├─PROVIDER: "RaspberryPi PWM Provider" {raspberrypi-pwm} <com.pi4j.plugin.raspberrypi.provider.pwm.RpiPwmProviderImpl> {com.pi4j.plugin.raspberrypi.provider.pwm.RpiPwmProviderImpl}
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: │ └─PROVIDER: "PiGpio PWM Provider" {pigpio-pwm} <com.pi4j.plugin.pigpio.provider.pwm.PiGpioPwmProviderImpl> {com.pi4j.plugin.pigpio.provider.pwm.PiGpioPwmProviderImpl}
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: ├─DIGITAL_OUTPUT: [2] <com.pi4j.io.gpio.digital.DigitalOutputProvider>
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: │ ├─PROVIDER: "PiGpio Digital Output (GPIO) Provider" {pigpio-digital-output} <com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalOutputProviderImpl> {com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalOutputProviderImpl}
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: │ └─PROVIDER: "RaspberryPi Digital Output (GPIO) Provider" {raspberrypi-digital-output} <com.pi4j.plugin.raspberrypi.provider.gpio.digital.RpiDigitalOutputProviderImpl> {com.pi4j.plugin.raspberrypi.provider.gpio.digital.RpiDigitalOutputProviderImpl}
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: ├─SPI: [2] <com.pi4j.io.spi.SpiProvider>
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: │ ├─PROVIDER: "PiGpio SPI Provider" {pigpio-spi} <com.pi4j.plugin.pigpio.provider.spi.PiGpioSpiProviderImpl> {com.pi4j.plugin.pigpio.provider.spi.PiGpioSpiProviderImpl}
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: │ └─PROVIDER: "RaspberryPi SPI Provider" {raspberrypi-spi} <com.pi4j.plugin.raspberrypi.provider.spi.RpiSpiProviderImpl> {com.pi4j.plugin.raspberrypi.provider.spi.RpiSpiProviderImpl}
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: ├─SERIAL: [2] <com.pi4j.io.serial.SerialProvider>
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: │ ├─PROVIDER: "PiGpio Serial Provider" {pigpio-serial} <com.pi4j.plugin.pigpio.provider.serial.PiGpioSerialProviderImpl> {com.pi4j.plugin.pigpio.provider.serial.PiGpioSerialProviderImpl}
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: │ └─PROVIDER: "RaspberryPi Serial Provider" {raspberrypi-serial} <com.pi4j.plugin.raspberrypi.provider.serial.RpiSerialProviderImpl> {com.pi4j.plugin.raspberrypi.provider.serial.RpiSerialProviderImpl}
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]: └─DIGITAL_INPUT: [2] <com.pi4j.io.gpio.digital.DigitalInputProvider>
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]:   ├─PROVIDER: "RaspberryPi Digital Input (GPIO) Provider" {raspberrypi-digital-input} <com.pi4j.plugin.raspberrypi.provider.gpio.digital.RpiDigitalInputProviderImpl> {com.pi4j.plugin.raspberrypi.provider.gpio.digital.RpiDigitalInputProviderImpl}
Mar 11 10:05:22 MediaParkPiBell pibell-service.jar[2215]:   └─PROVIDER: "PiGpio Digital Input (GPIO) Provider" {pigpio-digital-input} <com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalInputProviderImpl> {com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalInputProviderImpl}
Mar 11 10:05:49 MediaParkPiBell systemd[1]: session-6.scope: Succeeded.
```

Unfortunatley my Listener is still not triggered (I added log output as first statement there, also no breakpoint holds there).

**Comment by FDelporte** [comment]
From what I see here the output looks OK... `pigpio-digital-input` you use in the config of your input is loaded.

Can you also output the registry?
https://github.com/Pi4J/pi4j-example-minimal/blob/main/src/main/java/com/pi4j/example/MinimalExample.java#L133

**Comment by StefanSchubert** [comment]
I added the registry call, I looks like this now:

```
...
Mar 11 21:00:47 MediaParkPiBell pibell-service.jar[1674]: └─DIGITAL_INPUT: [2] <com.pi4j.io.gpio.digital.DigitalInputProvider>
Mar 11 21:00:47 MediaParkPiBell pibell-service.jar[1674]:   ├─PROVIDER: "PiGpio Digital Input (GPIO) Provider" {pigpio-digital-input} <com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalInputProviderImpl> {com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalInputProviderImpl}
Mar 11 21:00:47 MediaParkPiBell pibell-service.jar[1674]:   └─PROVIDER: "RaspberryPi Digital Input (GPIO) Provider" {raspberrypi-digital-input} <com.pi4j.plugin.raspberrypi.provider.gpio.digital.RpiDigitalInputProviderImpl> {com.pi4j.plugin.raspberrypi.provider.gpio.digital.RpiDigitalInputProviderImpl}
Mar 11 21:00:47 MediaParkPiBell pibell-service.jar[1674]: REGISTRY: [0] "I/O Registered Instances" <com.pi4j.registry.impl.DefaultRegistry>
Mar 11 21:01:22 MediaParkPiBell systemd[1]: session-5.scope: Succeeded.
```

Looks like there is someting missing?
I also saw from your vaadin example that you are running on java17 while I was on java11. To close this difference I choosed to install sdkman and I managed to fetch the 32bit version of java 17 for the pi as well:

```
pi@MediaParkPiBell:/var/pibell/logs $ java --version
openjdk 17.0.1 2021-10-19 LTS
OpenJDK Runtime Environment (build 17.0.1+12-LTS)
OpenJDK Server VM (build 17.0.1+12-LTS, mixed mode)

```
Project compiles now with java17 also. I had some troubles with the switch first, as the fat jar could not find java first after switching to sdkman. But setting the java_home also in the service.conf solved the problem with location java at service start.

**Reply by StefanSchubert** [reply]
Ah...I printed the registry info before doing this:

`  sensorInput = pi4jContext.create(noiceSensorConfig);
`
I will add a second call afterwards. And will post the results again. But now Sunday house duties are calling ;-)

**Reply by StefanSchubert** [reply]
Makes no differenece:

`
...
Mar 13 12:47:46 MediaParkPiBell pibell-service.jar[1917]: REGISTRY: [0] "I/O Registered Instances" <com.pi4j.registry.impl.DefaultRegistry>
Mar 13 12:47:46 MediaParkPiBell pibell-service.jar[1917]: REGISTRY: [0] "I/O Registered Instances" <com.pi4j.registry.impl.DefaultRegistry>
Mar 13 12:48:20 MediaParkPiBell systemd[1]: session-1.scope: Succeeded.`

Before and after the statement same effect. What would be the right place for a breakpoint in PI4J to see what's went wrong?

**Comment by StefanSchubert** [comment]
Since I'm tapping in the dark, I tried the one thing or the other, I used a static class to to the initialization, in your vaadin example you use a service. As far as I know a spring service could create some instances - that was why I used the static approach first, I will switch to an ApplicationScope Bean afterwards.

However I refactored me setup and now I saw 

```
2022-03-13 20:46:20,150 ERROR d.a.i.d.p.PiAdapter [main] #### Can't provision Input Pins!
java.lang.reflect.UndeclaredThrowableException: null
	at jdk.proxy2.$Proxy75.create(Unknown Source) ~[?:?]
	at com.pi4j.context.Context.create(Context.java:317) ~[pi4j-core-2.1.1.jar!/:?]
	at com.pi4j.internal.IOCreator.create(IOCreator.java:58) ~[pi4j-core-2.1.1.jar!/:?]
	at com.pi4j.internal.IOCreator.create(IOCreator.java:106) ~[pi4j-core-2.1.1.jar!/:?]
...
```

in my app logs. Which I haven't seen before.  As I assume that my systemd setup may cause problems, I will first see that it's working by invoking by using sudo.

Doing this the PrintInfo suddenly shows something in the registry (and the exception above vanishes):

```
PLATFORM: "RaspberryPi Platform" {raspberrypi} <com.pi4j.plugin.raspberrypi.platform.RaspberryPiPlatform> {Pi4J Platform for the RaspberryPi series of products.}
PROVIDERS: [12] "I/O Providers" <com.pi4j.provider.impl.DefaultProviders>
├─DIGITAL_OUTPUT: [2] <com.pi4j.io.gpio.digital.DigitalOutputProvider>
│ ├─PROVIDER: "RaspberryPi Digital Output (GPIO) Provider" {raspberrypi-digital-output} <com.pi4j.plugin.raspberrypi.provider.gpio.digital.RpiDigitalOutputProviderImpl> {com.pi4j.plugin.raspberrypi.provider.gpio.digital.RpiDigitalOutputProviderImpl}
│ └─PROVIDER: "PiGpio Digital Output (GPIO) Provider" {pigpio-digital-output} <com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalOutputProviderImpl> {com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalOutputProviderImpl}
├─DIGITAL_INPUT: [2] <com.pi4j.io.gpio.digital.DigitalInputProvider>
│ ├─PROVIDER: "RaspberryPi Digital Input (GPIO) Provider" {raspberrypi-digital-input} <com.pi4j.plugin.raspberrypi.provider.gpio.digital.RpiDigitalInputProviderImpl> {com.pi4j.plugin.raspberrypi.provider.gpio.digital.RpiDigitalInputProviderImpl}
│ └─PROVIDER: "PiGpio Digital Input (GPIO) Provider" {pigpio-digital-input} <com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalInputProviderImpl> {com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalInputProviderImpl}
├─SERIAL: [2] <com.pi4j.io.serial.SerialProvider>
│ ├─PROVIDER: "PiGpio Serial Provider" {pigpio-serial} <com.pi4j.plugin.pigpio.provider.serial.PiGpioSerialProviderImpl> {com.pi4j.plugin.pigpio.provider.serial.PiGpioSerialProviderImpl}
│ └─PROVIDER: "RaspberryPi Serial Provider" {raspberrypi-serial} <com.pi4j.plugin.raspberrypi.provider.serial.RpiSerialProviderImpl> {com.pi4j.plugin.raspberrypi.provider.serial.RpiSerialProviderImpl}
├─PWM: [2] <com.pi4j.io.pwm.PwmProvider>
│ ├─PROVIDER: "RaspberryPi PWM Provider" {raspberrypi-pwm} <com.pi4j.plugin.raspberrypi.provider.pwm.RpiPwmProviderImpl> {com.pi4j.plugin.raspberrypi.provider.pwm.RpiPwmProviderImpl}
│ └─PROVIDER: "PiGpio PWM Provider" {pigpio-pwm} <com.pi4j.plugin.pigpio.provider.pwm.PiGpioPwmProviderImpl> {com.pi4j.plugin.pigpio.provider.pwm.PiGpioPwmProviderImpl}
├─I2C: [2] <com.pi4j.io.i2c.I2CProvider>
│ ├─PROVIDER: "RaspberryPi I2C Provider" {raspberrypi-i2c} <com.pi4j.plugin.raspberrypi.provider.i2c.RpiI2CProviderImpl> {com.pi4j.plugin.raspberrypi.provider.i2c.RpiI2CProviderImpl}
│ └─PROVIDER: "PiGpio I2C Provider" {pigpio-i2c} <com.pi4j.plugin.pigpio.provider.i2c.PiGpioI2CProviderImpl> {com.pi4j.plugin.pigpio.provider.i2c.PiGpioI2CProviderImpl}
├─ANALOG_OUTPUT: [0] <com.pi4j.io.gpio.analog.AnalogOutputProvider>
├─SPI: [2] <com.pi4j.io.spi.SpiProvider>
│ ├─PROVIDER: "PiGpio SPI Provider" {pigpio-spi} <com.pi4j.plugin.pigpio.provider.spi.PiGpioSpiProviderImpl> {com.pi4j.plugin.pigpio.provider.spi.PiGpioSpiProviderImpl}
│ └─PROVIDER: "RaspberryPi SPI Provider" {raspberrypi-spi} <com.pi4j.plugin.raspberrypi.provider.spi.RpiSpiProviderImpl> {com.pi4j.plugin.raspberrypi.provider.spi.RpiSpiProviderImpl}
└─ANALOG_INPUT: [0] <com.pi4j.io.gpio.analog.AnalogInputProvider>
**REGISTRY: [1] "I/O Registered Instances" <com.pi4j.registry.impl.DefaultRegistry>
└─IO: "KY-038 Module" {BCM-#27} <com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalInput> {DoorBellSound Detection using KY-038 module from AZ-Delivery in threshold mode}**

```

I was almost ready to jump up from the couch and doing the hamster-dance, however my listener is still not being triggerd.
So I would say I'm a little step further, but I still seem to miss something crucial, that the eventhandling does not take place.

My Listener is Autowired, but this seem not to work in my setup:

```
2022-03-13 21:54:49,410 WARN d.a.i.d.p.PiAdapter [main] #### Couldn't attach a Listener on required digital Input PIN.
2022-03-13 21:54:49,412 WARN d.a.i.d.p.PiAdapter [main] +--> pinEventListener was null
```

Strange, but this is solvable...

**Comment by FDelporte** [comment]
just noticed this line
`.debounce(DEBOUNCE_DELAY_IN_SECS, TimeUnit.MILLISECONDS) // convert secs to millis`

the value and timeunit must match, so I guess you would need TimeUnit.SECONDS based on the name of your debounce value.

this is the Pi4J code you are calling from this

```
public DigitalInputConfigBuilder debounce(Long interval, TimeUnit units) {
        return debounce(units.toMicros(interval));
    }
```

**Reply by StefanSchubert** [reply]
So here is the next update:

- Fixed the TimeUnit Issue (thanks for the hint)
- I fixed my Autowire problem (problem was expecting the values wired during constructor call), fixed this through a lazy init step.
- Currently still running with the sudo to check if the fat jar is working (by just: sudo myFat.jar)
- Now Registry available as listed in previous post and this time Event Lister is added.

However no events seem to be calling the Listener. Hm, where in Pi4J is the event handling expected? I would like to set a breakpoint there to see why the events are not being propagated to the Listener.
